src/shared/logs-show.c

```diff
@@ -1595,21 +1611,36 @@ int add_matches_for_unit(sd_journal *j, const char *unit) {
         return r;
 }
 
-int add_matches_for_user_unit(sd_journal *j, const char *unit, uid_t uid) {
+int add_matches_for_user_unit_full(sd_journal *j, sd_bus *bus, const char *unit, uid_t uid, bool add_invocation) {
+        sd_id128_t id;
         int r;
 
         assert(j);
         assert(unit);
 
+        if (add_invocation) {
+                assert(bus);
+
+                r = unit_get_invocation_id(bus, unit, &id);
+                if (r < 0)
+                        return r;
+                if (r == 0) {
+                        log_debug("No invocation ID found for %s.", unit);
+                        add_invocation = false;
+                }
+        }
+
         (void) (
                 /* Look for messages from the user service itself */
                 (r = journal_add_match_pair(j, "_SYSTEMD_USER_UNIT", unit)) ||
                 (r = journal_add_matchf(j, "_UID="UID_FMT, uid)) ||
+                (r = add_invocation ? journal_add_match_pair(j, "_SYSTEMD_INVOCATION_ID", SD_ID128_TO_STRING(id)) : 0) ||
```

Comment:
hmm, matches are slow, hence I think we should reduce them as much as we can. Thing is, if we have an invocation ID, then we really don't need to track the UID/unt anymore, it should be the right one anyway.

Hence I think it makes even more sense to move this to its own helper function, because we really should only match against the invocation ID and that's it.

(invocation ids are always freshly generated, they should not be reused by other users/units, hence it should suffice to match against them)

Reply:
(hmm, so actually for the per-user invocation ID it might make sense to keep the _UID= part in, since the invocation id is not trusted then, hence it makes sense to limit the match with an additional trusted field, i.e. UID=)